Document `let_chains` again

est31 opened the pull request:
Mostly a reapply of #1179 which was reverted by #1251. The reapply is not trivial as in the meantime there has been #1561 and #1787, both of which caused merge conflicts. Also add some modifiers.

Tracking issue: https://github.com/rust-lang/rust/issues/53667
Stabilization report: https://github.com/rust-lang/rust/pull/132833

est31 committed:
Document let chains again

ehuss commented:
Some notes about temporaries and drop order:

TC's example playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=0ce26b5398487e95672adbf389ad946b

Niko's writeup: https://github.com/rust-lang/rust/pull/132833#issuecomment-2701981915

ehuss commented on src/expressions/if-expr.md:
```diff
@@ -1,35 +1,41 @@
+> &nbsp;&nbsp; _IfCondition_ ( && _IfCondition_ )*
+>
+> _IfCondition_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [_Expression_]<sub>_except struct expression_</sub>\
+> &nbsp;&nbsp; | `let` [_Pattern_] `=` [_Scrutinee_]
```
The expression cannot be a lazy-boolean, correct?

(_Scrutinee_ already says it doesn't allow struct. It's a little redundant to say that here, but maybe it helps with clarity?)

```suggestion
> &nbsp;&nbsp; &nbsp;&nbsp; [_Expression_]<sub>_except struct expression or lazy boolean expression_</sub>\
> &nbsp;&nbsp; | `let` [_Pattern_] `=` [_Scrutinee_]<sub>_except struct expression or lazy boolean expression_</sub>
>
> [^if-condition-2024]: Editions before 2024 only allow a single _IfCondition_.
```

ehuss commented on src/expressions/if-expr.md:
```diff
@@ -1,35 +1,41 @@
+The syntax of an `if` expression is a sequence of one or more condition operands separated by `&&`,
+followed by a consequent block, any number of `else if` conditions and blocks, and an optional trailing `else` block.
 
 r[expr.if.condition-bool]
```
Suggesting to rename this, since the condition can now have `let` in addition to boolean expressions.

```suggestion
r[expr.if.condition]
```

ehuss commented on src/expressions/if-expr.md:
```diff
@@ -1,35 +1,41 @@
+> &nbsp;&nbsp; (`else` ( [_BlockExpression_] | _IfExpression_ ) )<sup>\?</sup>
+>
+> _IfConditions_ :\
+> &nbsp;&nbsp; _IfCondition_ ( && _IfCondition_ )*
```
```suggestion
> &nbsp;&nbsp; _IfCondition_ ( && _IfCondition_ )*[^if-condition-2024]
```

ehuss reviewed (commented):
### Scopes
I think the rule in `scopes.md` labeled `names.scopes.pattern-bindings.if-let` needs to be updated. My rough understanding is that the scope of an `if let` binding extends into the *following* _IfCondition_s and the consequent block.

### Edition
At the bottom of `if-expr.md`, can you add something like:

```
r[expr.if.edition2024]
> **Edition differences**: Before the 2024 edition, let chains are not supported and only a single _IfCondition_ is allowed in an `if` expression.
```

I think it makes sense to be at the end of the "Chains of conditions" section? Also just want to double-check the accuracy here.

### Destructors and drop scopes

I'd like to see some updates to the destructors page to explain the drop scopes affected by this change. I'm uncertain exactly which changes to make, but I think roughly they are:

* `destructors.scope.desugaring` -- `if let` (now just `if`) can't be directly converted to a match expression, so I don't know how to phrase that.
* `destructors.scope.temporary.enclosing` -- `if let` should be rephrased somehow. Perhaps it could be phrased as just `if`, and somewhere say that multiple _IfCondition_s are translated to the equivalent nested `if` with single conditions?
    * Or, somehow word it in a similar way to how lazy boolean expressions are worded?
* `destructors.scope.temporary.edition2024` -- I'm uncertain if this needs rewording?

est31 commented:
Thanks for the review. Applied your suggestions and addressed the other review points as well.

> if let (now just if) can't be directly converted to a match expression, so I don't know how to phrase that.

When I read that sentence, I thought it's fine, because there is probably a way to express everything with nested match plus break from loops. But it's a bit underspecified, so I made it a little bit more explicit that it's not just a straightforward translation to `match`.

I made the new sentence suggest the new meaning a bit more clearly, but it's also now stating a bit of a tautology. 

> * `destructors.scope.temporary.edition2024` -- I'm uncertain if this needs rewording?

I think it's fine. One could swap out the `if let` with some longer wording, but it already feels complicated enough for me, I think.

ehuss commented:
@est31 Just checking if you saw the question about `names.scopes.pattern-bindings.if-let`, where I think the name resolution rules need updating? Is it correct that the bindings are in scope for the following conditions (and consequent)?

Looks like there is also a broken link error in `scopes.md`.

est31 commented:
Indeed the bindings are available for all following conditions in the chain (as well as the body). I will adjust the PR.

est31 commented on src/expressions/if-expr.md:
````diff
@@ -1,34 +1,40 @@
 ```grammar,expressions
 IfExpression ->
-    `if` Expression _except [StructExprStruct]_ BlockExpression
-    (`else` ( BlockExpression | IfExpression | IfLetExpression ) )?
+    `if` IfConditions BlockExpression
+    (`else` ( BlockExpression | IfExpression ) )?
+
+IfConditions -> IfCondition ( `&&` IfCondition )*
+
+IfCondition ->
+      Expression _except [StructExprStruct]_
+    | `let` Pattern `=` Scrutinee _except [LazyBooleanExpression]_
````
> That is, the exception isn't just `LazyBooleanExpression`; it's any boolean expression with a lower precedence than `&&`.

You are right that the exception isn't just that, and maybe it's misleading to list only `LazyBooleanExpression` because people might think that's an exhaustive list of exceptions.

From what I can see, the reference doesn't require the grammar to be "precedence fitting", i.e. only allow what's possible to express in text form following precedence rules. Instead, it puts the precedence into a table and assumes the grammar is parsed according to the precedence. I would say that table also applies here, as it's part of the larger "expression" family of things.

Because people might think it's exhaustive, we could maybe remove the exception altogether and replace it with a note instead?

> we're hoping for `Expression` to try to be `true || false && let () = ()`, notice that's impossible, and fail (i.e., rather than allowing `Expression` to be `true || false`, which would be wrong).

`||` has weaker precedence, so there isn't a top level && chain any more. On the other hand, `==` has stronger precedence.

The example with `||` is parsed to sth like:

```
If(
    BinOp(
        ||,
        true,
        BinOp(
            &&,
            false,
            Let((), ()),
        )
    ),
    Block([])
)
```

The example with `==` is parsed like:

```

If(
    BinOp(
        &&,
        true,
        BinOp(
            ==,
            false,
            Let((), ()),
        )
    ),
    Block([])
)
```

So in the `||` example, the BinOp with `&&` is actually not top level, its arms can't be `IfCondition`s. So its arms don't have a "permit to let". While in the `==` example, it has such a permission.

So I think this property is emergent from the grammar and the fact that it is parsed according to precedence.

est31 replied:
Dumping the AST, it makes sense that it stumbles across that.

First example is parsed like:
```
If(
    BinOp(
        &&,
        Let((), ()),
        Range(None, Some(0))
    ),
    Block([])
)
```
While the second example is parsed like:
```
If(
    Range(
        Some(0),
        Some(
            Let((), ())
        )
    ),
    Block([])
)
```

It's apparently how the parsing works for ranges, you can see similar effects with `_ = false && ..0;` vs `_ = 0.. && false`.

Not sure whether it's a bug or if it's derived from the reference, but I'd say it's existing (and mostly orthogonal) behaviour.

traviscross committed:
Use edition block syntax

Let's switch the let chains update to use our the markdown syntax for
edition differences.

traviscross committed:
Remove stray text

ehuss commented on src/expressions/if-expr.md:
```diff
@@ -148,27 +103,71 @@ if let E::X(n) | E::Y(n) = v {
-[_eRFCIfLetChain_]: https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#rollout-plan-and-transitioning-to-rust-2018
-[`match` expression]: match-expr.md
+r[expr.if.edition2024]
+> [!EDITION-2024]
+> Before the 2024 edition, let chains are not supported and only a single _IfCondition_ is allowed in an `if` expression.
```
IIUC, this will need to be duplicate in loop-expr.md for `while` loops, too, correct?
(That is, until that restriction is removed.)

est31 replied:
when I wrote https://github.com/rust-lang/rust/pull/140204#issuecomment-2824605396 I thought so too, but there is this paragraph in the `while let patterns` section:

> Multiple condition operands can be separated with `&&`.
> These have the same semantics and restrictions as [`if` condition chains].

I'd say that this one implies that while let chains are under the same edition specific restriction.

ehuss committed:
Update let chain grammar for minimum precedence

This adds some more exclusions to the let chain grammar to capture the
minimum precedence while parsing chains. Note that this is incomplete,
see https://github.com/rust-lang/reference/issues/1811. I didn't want to
block things on finalizing this.

I also reworked the `while` grammar to just reuse the grammar from `if`
to avoid the duplication.

ehuss committed:
Add some redirects for sections that were renamed

This adds some redirects for header fragments that have changed to
help external links go to the correct location.

ehuss reviewed (approved):
Thanks!

Rather than block on any more discussion on the grammar, I pushed a few modifications to start to express some of the precedence restriction, but it is incomplete and I filed https://github.com/rust-lang/reference/issues/1811 to track follow up work.
